### PR TITLE
Misc fixes for LOD syncing and upgrade Morango

### LINKED ISSUE #11524: Beta 8 Android syncing issues
## Observed behavior

After setting up a LOD on an Android device the syncing is working correctly for some time and then it completely stops regardless of how many times I restart the server and the app on the device. I'm providing the logs of 2 separate testing sessions on Android using the same learner and the server logs. Also tested syncing with a Mac and Windows LODs which were syncing correctly.

It seems that the issue described in https://github.com/learningequality/kolibri/issues/11485 is still valid as well.

## Expected behavior

Fully functioning sync with an Android LOD.

## Steps to reproduce the issue
1. Install the [Beta 8 build](https://github.com/learningequality/kolibri/releases/tag/v0.16.0-beta8).
2. Setup a server and an Android LOD device connected to the server.
3. Start assigning lessons/quizzes to the LOD and observe the sync status.

## Logs

https://drive.google.com/drive/folders/1TFxMYrFIj0BpjD3KdpdFHwlHsy9DnfUp?usp=drive_link

## Usage Details
Ubuntu 22  - Chrome, Android 13

### PR DESCRIPTION
**Pull request by bjester**
## Summary

- Updates LOD sync processing to never cancel the task but just enqueue and catch `JobRunning`
- Updates LOD sync processing to reset any queue records that are marked as `Syncing` since there should only ever be one job running that code at a time, and records with that status indicate it was terminated unexpectedly
- Adds check in cleanup hook that enqueues sync garbage collection to only occur if Kolibri is provisioned
- Adds defensive check to LOD syncing for unavailable network locations and backs off syncing in that scenario
- Upgrades morango to 0.6.19, fixing issue with premature cleanup of LOD sync sessions interfering with resumption
- Adds more unit tests for LOD syncing code

## References

Resolves https://github.com/learningequality/kolibri/issues/11524
Includes https://github.com/learningequality/morango/pull/205

## Reviewer guidance

Sync

----

## Testing checklist

- [X] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [X] Critical and brittle code paths are covered by unit tests

## PR process

- [X] PR has the correct target branch and milestone
- [X] PR has 'needs review' or 'work-in-progress' label
- [X] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

**Comment by radinamatic** [comment]
These are just logs after just one LoD import: it synced Ok for about 90 minutes, and then run into the usual wall 🤷🏽‍♀️ 

[Android13-db-and-logs.zip](https://github.com/learningequality/kolibri/files/13342227/Android13-db-and-logs.zip)
[Windows10-db-and-logs.zip](https://github.com/learningequality/kolibri/files/13342228/Windows10-db-and-logs.zip)

**Comment by pcenov** [comment]
Hi @bjester adding additional logs for 3 consecutively created Android LODs: https://drive.google.com/drive/folders/1__X6gsdgSh-s6UtdwvSk8qg-NiUG69FJ?usp=sharing

I'm still not able to pinpoint the exact cause, hoping the logs will give you a better picture of what's going on!

**Comment by radinamatic** [comment]
With the latest assets from yesterday, the first LoD (Android13) was syncing perfectly (user import, initial and new assignments download, added resources, etc.) for about 2 hours. But about 20 minutes after I synced another LoD (Android12, user created on the server's facility this time), the first one stopped, as in previous tests... 😞 

On the second LoD (Android12), the cards of class assignments appeared much faster on the Home screen (in less than a minute), but the resources did not download at all (but the status on both the Coach side and on LoD would state _Synced_): every time I would try to tap on the card to actually open the lesson or a quiz, it would forward me to the Library screen. Eventually, the second device also stopped syncing, less than an hour after the setup, and I never got to see and interact with any of the resources.    

[Android11-michc-db-logs.zip](https://github.com/learningequality/kolibri/files/13363929/Android11-michc-db-logs.zip), (this is actually Android 12, I forgot it went through an upgrade)
[Android13-peterb-db-logs.zip](https://github.com/learningequality/kolibri/files/13363931/Android13-peterb-db-logs.zip)
[Windows10-db-logs.zip](https://github.com/learningequality/kolibri/files/13363933/Windows10-db-logs.zip)

**Comment by radinamatic** [comment]
Since yesterday it took us some time with @rtibbles to figure out some subtleties of using `adb`, I decided to give it a go again today, trying to capture the time frame when a LoD stops syncing. 

1. I cleared the data from the 2 devices that have previously stopped syncing (learners `peterb` and `michc`, logs included in the previous post)
2. I imported the learner `neelar` on the Android 13 device and played with it for about an hour, with syncing working well
3. At hour mark I start importing the learner `marcg` on the Android 12 device, and (surprise, surprise), Android 13 device stops syncing right about then.
4. I fired the USB debugging immediately, and collected the `adb` output here: [Android13-neelar.log.txt](https://github.com/learningequality/kolibri/files/13366198/Android13-neelar.log.txt)
5. Android 12 device is currently still syncing, will add `adb` output logs if/when it stops.

**Comment by rtibbles** [comment]
I am seeing this sequence in the logs:
```
11-15 13:21:53.337 14650 14785 I WM-WorkerWrapper: Worker result SUCCESS for Work [ id=2f64f5bb-e9f6-4ea3-9f1c-b7c3d1c229d0, tags={ kolibri_job_type:kolibri.core.auth.tasks.soud_sync_processing, kolibri_task_id:50, org.learningequality.Kolibri.TaskworkerWorker } ]
11-15 13:21:53.345 14650 14785 I WM-WorkerWrapper: Setting status to enqueued for cfec18da-211c-4944-a715-5ba7b772329f
```

The first is the current soud_sync_processing task being successfully completed, the next is the enqueueing of the follow up task (now that its prerequisite has completed) - however, I'm never seeing that task get processed. Unless something very strange is happening with WorkManager, the only thing I can think of is if somehow we are setting the delay to a huge increment.

**Comment by radinamatic** [comment]
Just to confirm that even the very beginning of the learner import process on another LoD seems to interrupt the syncing of the LoD that was previously syncing correctly.

LoD with the learner `marcg` (Android 12 device) was syncing fine for a couple of hours, but as soon I started the process of setting up a new LoD, it stopped. I actually did not even finish the user import of the user `susanl` on the Android 9 tablet, just selected the server device, and the syncing on Android 12 device (with the learner `marcg`) was interrupted.

IMG_20231115_161135548

`adb` output here for Android 12: [Android12-marcg.log.txt](https://github.com/learningequality/kolibri/files/13367038/Android12-marcg.log.txt)

Android 9 tablet is currently still syncing...

**Comment by radinamatic** [comment]
Kolibri logs from the server and the 2 LoDs in this latest round of syncing.

[Windows10-server-db-logs.zip](https://github.com/learningequality/kolibri/files/13367100/Windows10-server-db-logs.zip)
[Android12-marcg-db-logs.zip](https://github.com/learningequality/kolibri/files/13367101/Android12-marcg-db-logs.zip)
[Android13-neelar-db-logs.zip](https://github.com/learningequality/kolibri/files/13367102/Android13-neelar-db-logs.zip)

**Comment by rtibbles** [comment]
This is the correlated log sequence to the logcat logs (noting that the Python logs are in UTC, while the logcat logs are in CET, hence the time discrepancy):

```
INFO 2023-11-15 12:21:42,155 __main__ Starting Kolibri task worker, for job 50
INFO 2023-11-15 12:21:42,273 kolibri.core.device.soud [user=820a1f6161c90a27fc98527090a07454] [server=58d91f1aff4f6f98b3f5f4def1e6167f] SoUD sync ready
INFO 2023-11-15 12:21:42,424 kolibri.core.discovery.utils.network.client Attempting connections to variations of the URL: http://192.168.1.126:8080/
INFO 2023-11-15 12:21:42,454 kolibri.core.discovery.utils.network.client Attempting connection to: http://192.168.1.126:8080/
INFO 2023-11-15 12:21:42,512 kolibri.core.discovery.utils.network.client Success! We connected to: http://192.168.1.126:8080/api/public/info/?v=3
INFO 2023-11-15 12:21:43,129 kolibri.core.auth.management.utils Remotely preparing data
INFO 2023-11-15 12:21:45,662 kolibri.core.auth.management.utils No records transferred
INFO 2023-11-15 12:21:47,340 kolibri.core.content.kolibri_plugin Processing content requests for synced dataset: bc5dc3336c78409bca3dd06a694ae2de
INFO 2023-11-15 12:21:47,342 kolibri.core.content.utils.content_request Processing new content assignment requests
INFO 2023-11-15 12:21:47,351 kolibri.core.content.utils.content_request Processing new content removal requests
INFO 2023-11-15 12:21:47,360 kolibri.core.content.kolibri_plugin Completed content requests for synced dataset: bc5dc3336c78409bca3dd06a694ae2de
INFO 2023-11-15 12:21:47,617 __main__ Starting Kolibri task worker, for job 175d3c23e4504598b4701afe72408aaf
INFO 2023-11-15 12:21:47,694 __main__ Ending Kolibri task worker, for job 175d3c23e4504598b4701afe72408aaf
INFO 2023-11-15 12:21:48,113 __main__ Starting Kolibri task worker, for job dfeae49f27414d41bbfffe89bb842398
INFO 2023-11-15 12:21:48,475 __main__ Ending Kolibri task worker, for job dfeae49f27414d41bbfffe89bb842398
INFO 2023-11-15 12:21:48,501 kolibri.core.auth.management.utils Locally preparing data to send
INFO 2023-11-15 12:21:50,724 kolibri.core.auth.management.utils No records transferred
INFO 2023-11-15 12:21:53,259 __main__ Ending Kolibri task worker, for job 50
```
So, this shows the same story of the task successfully completing and the follow up being enqueued, but it then never happens.

**Comment by pcenov** [comment]
Here are my logs from today: https://drive.google.com/drive/folders/1KFFHncnxAsezcemNOxIDr1Ti0qMv2hkZ?usp=sharing
The AndroidLOD1 stopped syncing after several hours of working just fine. Then I created another AndroidLOD2 + Ubuntu and Windows LODs which have been syncing correctly while working together... Quite puzzling!

**Comment by radinamatic** [comment]
Promising results with the latest asset on a non-Xiaomi device: leaving the Kolibri app in the background did not interrupt syncing, and it was constant after various intervals of _backgrounding_. 👏🏽 

After that I left both devices _for the night_ (6-7 hours approximately) without closing the app. Upon resume the server reported last sync 7 hours ago, and the app _Not recently synced_ . I re-engaged a resource but the sync was not resuming. Only after the server restart was the sync resumed and the most recent interactions recorded in class activity.

[Windows10-db-logs.zip](https://github.com/learningequality/kolibri/files/13378114/Windows10-db-logs.zip)
[android12-rayb.log](https://github.com/learningequality/kolibri/files/13378125/android12-rayb.log)

**Comment by pcenov** [comment]
I first started with testing only with 1 Android LOD - unfortunately it did stop syncing after several hours of working just fine and without an obvious reason to do so.
I then set up another Android LOD with a couple of imported learners, a LOD in an emulator in Android Studio and one Ubuntu LOD all simultaneously syncing without observing any issues - really impressive! Here are all of the logs from today: https://drive.google.com/drive/folders/1VW0ULon0WHPhhR3nUddTx0DNKtTHyYNK?usp=sharing

File: kolibri/core/device/soud.py [hunk 1]
```diff
@@ -332,6 +332,15 @@ def execute_syncs():
     """
     Core SoUD sync processing logic that processes any syncs that
     """
+    # since there should only ever be one processing job running at a time, if we encounter any in
+    # the queue that are marked as syncing, we should reset their status to pending because it must
+    # mean that the previous job was terminated unexpectedly
+    SyncQueue.objects.filter(status=SyncQueueStatus.Syncing,).update(
```
Comment (rtibbles) [inline kolibri/core/device/soud.py:338]: :+1:

**Review by rtibbles — APPROVED** [review]
All makes sense to me.

I think the only other thing we can do on the Kolibri side here is making our job management more robust to job runner failure (which would be useful for both the Android App, and Cloud installations). Some sort of 'last updated' for a job or somesuch might be helpful to give us a heuristic to be able to reset tasks that report as running but are not really.

**Comment by pcenov** [comment]
@bjester here are the logs for an Android LOD that stopped syncing again after about 2 hours of syncing correctly: https://drive.google.com/drive/folders/1kfOwOcq7wrJhVKtXO49ykUx3Din4sJtL
I see that this PR is already merged but since there is still an issue with the syncing stopping (using the latest assets) this should be further investigated.

**Comment by bjester** [comment]
> @bjester here are the logs for an Android LOD that stopped syncing again after about 2 hours of syncing correctly: https://drive.google.com/drive/folders/1kfOwOcq7wrJhVKtXO49ykUx3Din4sJtL I see that this PR is already merged but since there is still an issue with the syncing stopping (using the latest assets) this should be further investigated.

@pcenov Richard noticed this in the Android logs
> NotificationService: Suppressing notification from package org.learningequality.Kolibri by user request.

Did you perhaps silence or disable the Kolibri notifications?

**Comment by pcenov** [comment]
@bjester yes, _think_ I had the notifications disabled on this particular Samsung device. But looking closely at this it can also be an issue on Samsung's end as I don't actually see an option to turn the notifications back on and could be caused by a recent update to Samsung's One UI 5.1 (Android 13).

**Pull request merged**